[aquaimpact] iP

aquaimpact opened the pull request:
# 🛡️ Aegis Task Manager Bot

_Aegis_ is a **command-line task management chatbot** built in **Java**, designed to help users manage their tasks efficiently. It supports adding, deleting, listing, and organizing tasks such as **Todos, Deadlines, and Events** with an easy-to-use command system.

---

## ✨ Features

✅ **Task Management:** Add, delete, and list various tasks (_Deadlines_, _Events_, _Todos_)  
✅ **Mark & Unmark Tasks:** Easily mark tasks as done or not done  
✅ **Sorting & Listing:** View all tasks or sort them by upcoming due dates  
✅ **Search Functionality:** Find tasks by name  
✅ **Error Handling:** Provides feedback for invalid commands  

---

## 📜 Commands

### 📝 Task Management

#### ✅ Mark a Task as Done  
- **Command:** `mark <task number>`  
- **Example:** `mark 2` (Marks task 2 as completed)  
- **Note:** `<task number>` must be greater than 0.  

#### ❌ Unmark a Task as Not Done  
- **Command:** `unmark <task number>`  
- **Example:** `unmark 2` (Marks task 2 as not completed)  
- **Note:** `<task number>` must be greater than 0.  

#### ➕ Add a Task  
- **Command (Todo):** `todo <task name>`  
  - Example: `todo Read book`  
- **Command (Deadline):** `deadline <task name> /by <date & time>`  
  - Example: `deadline Submit report /by 02/15/2025 18:00`  
- **Command (Event):** `event <task name> /from <date time> /to <date time>`  
  - Example: `event Team meeting /from 02/20/2025 14:00 /to 02/20/2025 16:00`  
- **Note:** Date format must be **MM/DD/YYYY HH:mm**.  

#### 🗑️ Delete a Task  
- **Command:** `delete <task number>`  
- **Example:** `delete 3` (Deletes task 3)  
- **Note:** `<task number>` must be greater than 0.  

---

### 📂 Listing Tasks

#### 📜 List All Tasks  
- **Command:** `list`  
- **Description:** Displays all tasks.  

#### 📆 List Tasks by Due Dates  
- **Command:** `duedates`  
- **Description:** Lists only _Deadlines_ and _Events_, sorted by the earliest due date first.  

---

## 🚀 Installation

1. **Download the latest Aegis JAR file** [here](https://github.com/aquaimpact/ip/releases/tag/A-Jar)  
2. **Place it in a folder of your choice**  
3. **Run it with the following command:**  `java -jar aegis.jar`
4. Start managing your tasks efficiently! 🎯  

---

## 📌 Current Progress :clipboard:

- [x] Implement core chatbot functionality (Add, List, Delete tasks)  
- [x] Refactor using **Object-Oriented Programming (OOP)**  
- [ ] Develop a **Graphical User Interface (GUI)** using JavaFX  

---
Here is where your save file is saved at!
```java
public static void main(String[] args) {
        new Aegis("./save.txt").run();
}
```

> _"A goal without a plan is just a wish."_ – Antoine de Saint-Exupéry  

📌 **Aegis: Your reliable task manager, built to shield your productivity!**

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

aquaimpact committed:
Rename main class

aquaimpact committed:
Add greeting

aquaimpact committed:
Add support for echo

aquaimpact committed:
Customise chatbot response

aquaimpact committed:
Fixed bugs in echo

aquaimpact committed:
Add support for list and add functions

aquaimpact committed:
Customise chatbot output

aquaimpact committed:
Customise chatbot output

aquaimpact committed:
Add Task.java

aquaimpact committed:
Update Task

aquaimpact committed:
Updated Task

aquaimpact committed:
Add support for mark and unmark

aquaimpact committed:
Customise chatbot output

aquaimpact committed:
Add Deadline, Event and Todo classes

aquaimpact committed:
Update Todo with methods

aquaimpact committed:
Add support for todo function

aquaimpact committed:
Optimise Task methods

aquaimpact committed:
Standardise variable names

aquaimpact committed:
Update Deadline methods

aquaimpact committed:
Add support for deadline

aquaimpact committed:
Update Event methods

aquaimpact committed:
Add support for event

aquaimpact committed:
Update text-ui-test

aquaimpact committed:
Update string formatting

aquaimpact committed:
Create support for invalid commands and inputs

aquaimpact committed:
Refector printing code

aquaimpact committed:
Add checks for UI testing

aquaimpact committed:
Add support for deleting tasks from the list

aquaimpact committed:
Add checks for UI testing

aquaimpact committed:
Convert command checks to enums and add javadocs and documentation

aquaimpact committed:
Add support for saving tasks onto the hard drive

aquaimpact committed:
Change return type for methods in FileSave class

aquaimpact committed:
Merge branch 'branch-Level-7'

aquaimpact committed:
Add support for storing deadlines

aquaimpact committed:
Display stored and showed date formats differently

aquaimpact committed:
Add support for ordering duedates based on urgency

aquaimpact committed:
Merge branch 'branch-Level-8'

aquaimpact committed:
Seperate command into CommandParser

aquaimpact committed:
Add TaskList to store tasks and UIManager to handle user interactions

aquaimpact committed:
Edit FileSave to change static methods to class methods

aquaimpact committed:
Shift the commands into a class

aquaimpact committed:
Modify Aegis to make use of the classes

aquaimpact committed:
Create classes for the individual commands

aquaimpact committed:
Seperate commands from the implementations

aquaimpact committed:
Merge branch 'branch-A-MoreOOP'

aquaimpact committed:
Seperate files into multiple packages

aquaimpact committed:
Add import statements for the Command class

aquaimpact committed:
Refector codes in the Command subclasses

aquaimpact committed:
Refector Aegis

aquaimpact committed:
Refactor test files

aquaimpact committed:
Merge branch 'branch-A-Packages'

aquaimpact committed:
Merge branch 'add-gradle-support'

aquaimpact committed:
Automate project build using Gradle

aquaimpact committed:
Add JUnit tests for AddCommand and DeleteCommand

aquaimpact committed:
Change save file location

aquaimpact committed:
Add JavaDoc to all the commands

aquaimpact committed:
Add JavaDoc for all exceptions

aquaimpact committed:
Add JavaDoc to CommandParser

aquaimpact committed:
Add JavaDoc to FileSave

aquaimpact committed:
Add JavaDoc to each task type and fix bugs

aquaimpact committed:
Add JavaDoc to UIManager

aquaimpact committed:
Reformat

aquaimpact committed:
Merge branch 'branch-A-JavaDoc'

aquaimpact committed:
Change order of imports

aquaimpact committed:
Tweak code

aquaimpact committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/aegis/storage/FileSave.java
#	src/main/java/aegis/task/Task.java
#	src/main/java/aegis/task/TaskList.java

aquaimpact committed:
Add support for finding tasks based on keywords

aquaimpact committed:
Add JavaDoc for FindCommand and supporting methods

aquaimpact committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/aegis/parser/CommandParser.java

aquaimpact committed:
Remove unused libraries

MrAsparag00se commented on src/main/java/Task.java:
```diff
@@ -10,7 +10,7 @@ public String getTaskName() {
-        return isComplete ? "X" : "";
+        return isComplete ? "X" : " ";
     }
     public void markAsDone() {
```
If markAsDone() returns true/false, would it be better if the method type is ```boolean``` instead?

MrAsparag00se commented on src/main/java/Aegis.java:
```diff
@@ -74,15 +74,18 @@ public static void main(String[] args) {
+                String[] res = input.split(" /by ");
+                String taskName = res[0].substring(9).trim();
+                String by = res[1].trim();
+                Task deadlineItm = new Deadline(taskName, by);
```
what does Itm stand for? Does not seem immediately obvious 🤔

MrAsparag00se commented on src/main/java/Deadline.java:
```diff
@@ -37,4 +37,14 @@ public String toString() {
     }
+
+    @Override
+    public int compareTo(Object o) {
```
Nice that you added a formatter to the list of tasks- did not think of doing that.

MrAsparag00se commented on src/main/java/Aegis.java:
```diff
@@ -15,220 +15,54 @@
-                if(inputArray.length == 1) throw new TaskInputException("You did not specify which task to mark done!");
-                if(tasks.isEmpty()) throw new TaskInputException("No task available to mark!");
+    private TaskList tasks;
+    private FileSave fs;
```
Should it be FileSave or SaveFile?

MrAsparag00se commented on src/main/java/DeleteClass.java:
Do you really need one class per command? Might that be over abstraction?

MrAsparag00se commented on src/main/java/Aegis.java:
```diff
@@ -44,12 +44,13 @@ public void run() {
+                Command c = CommandParser.parse(input);
+                c.execute(tasks, fs);
+                isExit = c.isExit();
             } catch (TaskInputException | CommandException e) {
```
Do you mean to type ```||``` instead?

MrAsparag00se commented on src/main/java/aegis/task/TaskList.java:
```diff
@@ -129,4 +131,18 @@ public ArrayList<Task> getTasks() {
+    public TaskList searchByName(String searchStr) {
+        ArrayList<Task> searchResults = new ArrayList<>(tasks.stream()
+                .filter(task -> task.getTaskName().toLowerCase().contains(searchStr.toLowerCase()))
+                .collect(Collectors.toCollection(ArrayList::new)));
```
Doesn't really matter to me, but I think use of collectors are not best practice. Not too sure.

MrAsparag00se reviewed (commented):
Overall LGTM. Seems like you are ahead wrt checkstyle and testing too. Keep it up!

aquaimpact committed:
Split lines into shorter parts

aquaimpact committed:
Merge branch 'branch-A-CheckStyle'

aquaimpact committed:
Refactor code to accomodate JavaFX

aquaimpact committed:
Merge branch 'branch-Level-10'

Victoria281 commented on src/main/java/aegis/parser/CommandParser.java:
```diff
@@ -0,0 +1,139 @@
+     * @throws CommandException If the command is invalid or unrecognized.
+     */
+    private static CommandType determineCommandType(String input) throws CommandException {
+        if (input.matches(".*\\bbye\\b.*")) {
+            return CommandType.BYE;
+        } else if (input.matches(".*\\blist\\b.*")) {
+            return CommandType.LIST;
+        } else if (input.matches(".*\\bmark\\b.*")) {
+            return CommandType.MARK;
+        } else if (input.matches(".*\\bunmark\\b.*")) {
+            return CommandType.UNMARK;
+        } else if (input.matches(".*\\bdelete\\b.*")) {
+            return CommandType.DELETE;
+        } else if (input.matches(".*\\btodo\\b.*")) {
+            return CommandType.TODO;
+        } else if (input.matches(".*\\bdeadline\\b.*")) {
+            return CommandType.DEADLINE;
+        } else if (input.matches(".*\\bevent\\b.*")) {
+            return CommandType.EVENT;
+        } else if (input.matches(".*\\bduedates\\b.*")) {
+            return CommandType.DUEDATES;
+        } else if (input.matches(".*\\bfind\\b.*")) {
+            return CommandType.FIND;
+        } else {
+            throw new CommandException(input);
+        }
```
I like that you're using regex for command checking! It makes the commands more flexible.

Victoria281 commented on src/main/java/aegis/parser/CommandParser.java:
```diff
@@ -0,0 +1,139 @@
+     */
+    public static Command parse(String input) throws TaskInputException, DateTimeParseException, CommandException {
+        String[] inputArray = input.split(" ");
+        CommandType ct = determineCommandType(input);
```
Would it be better to have a name for ct?

Victoria281 commented on src/main/java/aegis/storage/FileSave.java:
```diff
@@ -0,0 +1,127 @@
+        File f = new File(filePath);
+        Scanner s = new Scanner(f);
+
+        ArrayList<Task> arr = new ArrayList<>();
```
Would it be better to have a clearer name for this arr?

Victoria281 commented on src/main/java/aegis/storage/FileSave.java:
```diff
@@ -0,0 +1,127 @@
+        ArrayList<Task> arr = new ArrayList<>();
+
+        while (s.hasNext()) {
+            String[] taskData = s.nextLine().split("\\|\\|");
```
Perhaps put this in plural as it is a multi-valued variable?

Victoria281 commented on src/main/java/aegis/exception/FileSavingException.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Represents an exception that occurs during file-saving operations.
+ */
+public class FileSavingException extends AegisException {
```
Would FileSaveException be a clearer name here? 'Saving' is more of a verb, whereas class names typically follow a noun-based convention

Victoria281 reviewed (commented):
LGTM
Some cosmetic changes requested

aquaimpact committed:
Add assertions to Aegis class and remove unused import statements

- Added assertions to validate non-null inputs, ensure proper
initialization of TaskList, and check command execution results.
- Removed unused import statements to improve code clarity.
- Enhances robustness and prevents potential null pointer exceptions.

aquaimpact committed:
Add assertions to CommandParser for input validation

aquaimpact committed:
Merge pull request #1 from aquaimpact/branch-A-Assertions

Add Assertions and Clean Up Command Parsing

aquaimpact committed:
Remove unnecessary code and unused libraries

Removed unused import statements to improve code readability and
maintainability.

Eliminated redundant code to enhance efficiency and reduce clutter.

aquaimpact committed:
Add java docs and remove unused variables

This commit adds Javadoc comments to improve code documentation
and maintainability.
Unused variables have also been removed to enhance code clarity and
prevent potential confusion.

aquaimpact committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/aegis/Aegis.java
#	src/main/java/aegis/parser/CommandParser.java

aquaimpact committed:
Merge pull request #2 from aquaimpact/branch-A-CodeQuality

Improve Code Quality by Removing Unused Imports and Refactoring Code

aquaimpact committed:
Add gradle.yml

aquaimpact committed:
Add the ability to recognize and deal with duplicate items

aquaimpact committed:
Merge branch 'branch-C-DetectDuplicates'

aquaimpact committed:
Update documentation in README.md

aquaimpact committed:
Update aegis jar file download link

aquaimpact committed:
Extract task formatter for deadline and event tasks in command parser

aquaimpact committed:
Bug fixes for deadline and event comparison for duedates function

aquaimpact committed:
Add JUnit tests for DueDatesCommand

aquaimpact committed:
Update file saving flow

aquaimpact committed:
Refactor quit function

aquaimpact committed:
Remove unused methods and refactor file save code
